Support F-Client and replace non-extended extension messages

heinrich5991 opened the pull request:
This continues what #3439 started, and also replaces non-extended messages that are DDRace extensions with extended messages. The DDNet server still sends the old message to old clients and the DDNet client still understands the old messages.

Supersedes #3439.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

fokkonaut committed:
Support for F-Client

heinrich5991 committed:
Remove F-Client-specific messages

The PLAYERFLAG_AIM can now be set directly on DDNet servers via the 0.7
bridge and will be sent to clients identifying themselves as DDNet
clients.

Add a new extended message that mirror the old `Sv_TeamsState`.

heinrich5991 committed:
Replace remaining non-extended extension messages

Add backward compatibility stuff, but this allows newer clients to only
care for new messages.

heinrich5991 commented:
What do you think? @fokkonaut @Learath2

fokkonaut commented:
Exactly what I tried to avoid: Using the normal playerflags. 0.7 is still in development, and a new playerflag can always come which would break compability. Also this means I need to change the handling of unpacking the playerflags in the client, because they will fail on unpacking them.

fokkonaut commented:
Also the netobj handler 7 of ddnet server cannot unpack the playerflags.

fokkonaut commented:
Exactly what I tried to avoid: Using the normal playerflags. 0.7 is still in development, and a new playerflag can always come which would break compability.

Learath2 commented:
Is `Sv_DDRaceTime` used for anything on the client? Do we really need to keep support for prehistoric clients? If yes maybe we can just remove the msg for new clients all together?

I like the change, a rudimentary look suggests it's fine, but I need to give it a more thorough read.

heinrich5991 commented:
Yes, `Sv_DDRaceTime` is used in the client.

fokkonaut commented on src/game/server/player.cpp:
```diff
@@ -353,6 +355,8 @@ void CPlayer::Snap(int SnappingClient)
 			return;
 
 		pPlayerInfo->m_PlayerFlags = PlayerFlags_SixToSeven(m_PlayerFlags);
+		if(ClientVersion >= VERSION_DDRACE && (m_PlayerFlags & PLAYERFLAG_AIM))
```
This needs to check for the client version of the snapping client, not the own one.

heinrich5991 replied:
Done, I think.

fokkonaut commented:
I adapted to these changes and on my own server it works, just not correctly with this commit applied to ddnet servers.
F-Client commits: https://github.com/fokkonaut/F-Client/commit/a972c73d2404240ffe57737d7dabc2db03fe1b0f, https://github.com/fokkonaut/F-Client/commit/8ecd8c7f0ab4b41f16d165bc5ec9b9e87578012c
Also I noticed projectiles are invisible due to the extra info thats not used in 0.7 and therefore breaks the projectile. So we would need a check there too. I can receive teams, but not aimlines, seems like the ddnet server is not getting the clientver, in my mod it does get it.

fokkonaut commented:
Ah, I found why the hookline wont work. F-Client is not sending it because you need to increase the `SERVERCAP_CURVERSION` to 2 in protocol_ex.h. But what do we do about the projectile check? Could you add a !IsSixup there? (I hope it wont break anything, because F-Client is sending the latest ddnet version right now)

heinrich5991 commented:
> Ah, I found why the hookline wont work. F-Client is not sending it because you need to increase the `SERVERCAP_CURVERSION` to 2 in protocol_ex.h.

Done. Thanks for figuring that out.

> But what do we do about the projectile check? Could you add a !IsSixup there? (I hope it wont break anything, because F-Client is sending the latest ddnet version right now)

I don't see a difference between the 0.6 and 0.7 projectile netobj. Can you point me to the difference or where I should put that `!IsSixup` check?

fokkonaut commented:
https://github.com/ddnet/ddnet/blob/5d45f061fabf0866a9e8e6d5e59d542d3188ca78/src/game/server/entities/projectile.cpp#L325

Here. I hope there are no other cases in which the DDNet client version nr is passing F-Client through some stuff it doesnt have. We would need ClientCaps for the server :D

heinrich5991 committed:
Replace projectile hack with a new extended snapshot object

Still send the old hack to old clients, still accept the old hack from
old servers. This is so F-Client can support DDNet servers without such
hacks.

heinrich5991 commented:
@fokkonaut What do you think? I can PR support for DDNETPROJECTILE to F-Client if you want.

fokkonaut commented:
Sure.

Learath2 commented on src/game/server/entities/projectile.cpp:
```diff
@@ -319,13 +320,27 @@ void CProjectile::Snap(int SnappingClient)
+		{
+			return;
+		}
+		mem_copy(pProj, &DDNetProjectile, sizeof(DDNetProjectile));
```
This assumes the legacy and the new projectile message are laid out the same way in memory. Is this a fair assumption to make now? Is it guaranteed to stay a fair assumption in the future?

At least we should document this coupling somewhere

heinrich5991 replied:
I added documentation that it needs to have the same layout as `Projectile`. We already assume that we can read the structs as `int` arrays in the snapshot code.

Learath2 reviewed (commented):
We seem to be sending both the legacy and the new message to clients that seem new enough to receive the new message. Isn't that a bit wasteful (though convenient)?

heinrich5991 commented:
> We seem to be sending both the legacy and the new message to clients that seem new enough to receive the new message. Isn't that a bit wasteful (though convenient)?

I intended to always send the new message. And when the client is old, also send the legacy message it understands. This wastes network traffic for old clients, but eventually™ everyone upgrades to a client that is new enough.

Learath2 commented:
LGTM

bors r+
